Updated filemanager plugin to work with Micro editor version 2 plugin api change

tommyshem opened the pull request:
Note if this pull request gets in, the release will need making as version v4.0.0 due to all the changes and is not backwards compatible.

tommyshem committed:
convert microV2 api messenger: AddLog to micro.Log

tommyshem committed:
convert microv2 messenger: to  micro.InfoBar()

tommyshem committed:
change version to 4 as microV2 as breaking changes

tommyshem committed:
add imports needed for new microV2 api

tommyshem committed:
V2  convert read and write Options
changed GetOption to config.GetGlobalOption

Changed AddOption to config.RegisterCommonOption

Put the dots in as this is the new format for reading options

tommyshem committed:
micro v2 convert CurView to micro.CurPanel()

tommyshem committed:
micro v2 convert JoinPaths to filepath.Jin

tommyshem committed:
micro v2 convert WorkingDirectory to os.Getwd

tommyshem committed:
micro v2 convert .Buf.Cursor to .Cursor

tommyshem committed:
make glib_os global and remove all local calls
Using the import("os")

tommyshem committed:
micro v2 api RunShellCommand to shell.RunCommand

tommyshem committed:
golib_filepath made global and local ones removed

tommyshem committed:
micro v2 api DirectroyName to golib_filepath.Dir

tommyshem committed:
micro v2 api SetLocalOption to Buf:SetOption

tommyshem committed:
micro v2 api messenger:prompt changed to
 micro.InfoBar():YNPrompt

tommyshem committed:
micro api messenger:AddLog to micro.Log

tommyshem committed:
micro v2 api MakeCommand to config.MakeCommand

tommyshem committed:
micro v2 api messenger:Reset > micro.InfoBar:Reset

tommyshem committed:
missed os to golib_os changed

tommyshem committed:
put . back in which must have been removed
from the last commit

tommyshem committed:
remove the filemanager. from the MakeCommand
and also the " to match micro v2 api

tommyshem committed:
microv2 api AddRuntimeFile > config.AddRuntimeFile
add config.RTSyntax

tommyshem committed:
add missing . as must have got removed

tommyshem committed:
micro v2 api NewBuffer to buffer.NewBuffer

tommyshem committed:
micro v2 api NewBuffer commands flags changed
do not use int but bool

tommyshem committed:
micro v2 api view.Width >  GetView().Width

tommyshem committed:
change variable tree_view to tree_BufPane
as microv2 passes back a bufpane now instead of a view

tommyshem committed:
change . to : on the ResizePane as it is function

tommyshem committed:
micro v2 api Type > Buf.Type

tommyshem committed:
remove reset as not working in micro v2

tommyshem committed:
micro v2 api :insert to .EventHandler:Insert

tommyshem committed:
micro v2 api Buf.remove to Buf.EventHandler:Remove

tommyshem committed:
micro v2 api Loc to buffer.Loc

tommyshem committed:
micro v2 api tabs changed
tabs[curTab + 1]:Resize()
tree_BufPane:Tab():Resize()

tommyshem committed:
micro v2 callback function added

tommyshem committed:
micro v2 callback function preStartOfText missing

tommyshem committed:
micro v2 callback function preSelectToStartOfText
added

tommyshem committed:
add filemanager V4 and changed version order

tommyshem committed:
changed view variable to bufpane
As it is a bufpane object and not a view object

tommyshem committed:
micro v2 api remove false parameters not used

tommyshem committed:
add help file. Can  be displayed  in the editor

tommyshem committed:
micro v2 api command passes in bufpane
also the args passed in, is a string array instead of a string.

tommyshem committed:
update help file for micro editor

tommyshem committed:
update install instructions for version 2

sum01 commented on filemanager.lua:
```diff
@@ -802,68 +799,87 @@ local function create_filedir(filedir_name, make_dir)
+-- parse_args will check if the passed in string array is empty
+-- @param args string array[]
+-- @return nil or string [nil if the string array is empty]
+function parse_args(args)
+	local return_string = nil
+	if #args == 0 then
+		-- args is empty
+	else
+	return_string = args[1]
+	end
+	return return_string
+end
```
```suggestion
local function parse_args(args)
	if args == nil then
		micro.Log("parse_args given no args")
		return nil
	end
	local args_t = type(args)
	if args_t ~= "table" then
		if args_t == "string" then
			return args
		else
			micro.Log("parse_args not given a string or table type")
			return nil
		end
	end
	return #args > 0 and args[1] or nil
end
```

tommyshem replied:
I tried the above change and it passes back a userdata as type and it does not work. So left as is, as it is working and unless micro editor changes the api to the command parameter passing it will not need error checking.

sum01 reviewed (changes requested):
Why did the install instructions need to change? If the plugin is still in the official plugin repo, nothing should need to change to be able to download it.

---

I think the way you changed all the non-local functions will break the API for the user's `bindings.json` as previously none of those functions took a buffer as an arg.

I guess breaking API is fine, but I don't see a clean way for users to bind to it anymore, short of just passing `nil` as the first arg for every binding.

I'm assuming you did this because the new Micro commands require passing the buffer, but I'd suggest making some sort of micro-to-filemanager inbetween local functions to get around this. This would have the side-effect of not breaking API.

Something like
```lua
local function rename_at_cursor_internal(buf, args)
  if(buf ~= tree_buf) then
    micro.Log("Can't perform rename_at_cursor when the cursor isn't in the filemanager buffer.")
    return
  end 
  rename_at_cursor(parse_args(args))
end
```
that a Micro command is bound to, for example.

Note that if you do this & API doesn't break, then the major version shouldn't change (as per Semver). This would mean (assuming no new features added) that the new plugin version would be `3.4.1` since this would classify as a "bug fix".

I say "assuming no new features added" because I haven't dug through every bit of code to check yet if there were new features or not.

---

I'm also midly curious if some of the old workarounds (like for `cd` or cursor down) are still needed? If it isn't broken then I guess it isn't important though.

---

Lastly, you didn't update most of the comments where code changed. That should probably be fixed, either by updating comments so they make sense, or just deleting them entirely.

tommyshem commented:
The internal commands work with out passing the bufpane as micro auto sends it.

Most of the commands can not be keybind as they need an argument to be passed.

Commands below

eg
remove filename
touch filename
mkdir dirname
rename filename

The only 2 that can be keybind is `tree` and `rm` unless I am missing something here ?

And they have not been changed.

So no api change to the filemanager plugin but mass changes to the micro editor plugin calls api.

That is why I changed to version 4 as this plugin only works on micro editor version 2, and due to micro editor went from v1 to v2 I followed suit.

Also the config settings have changed as micro now uses `pluginname`.`settingname` as the name of the setting. Note the dot is inserted by micro editor.

I changed the install instructions as it needs more testing or if people are willing to try it. As it might cause data loss. Could do with putting beta on it until more testing can be done.

I hope I have given you enough information so you can make your mind up about this pull request.

I can change the comments on this pull request or in a new one, as this pull request is getting large.

Please let me know what you think.

sum01 commented:
Seems fine then. I guess micro has just changed a lot since v1, and I haven't kept up with it.

tommyshem commented:
Do you want me to update the comments in this pull request or create a new pull request ?

sum01 commented:
It's fine to keep it in this PR

tommyshem committed:
micro v2 change options in docs to dots
micro v2 uses plugin.option as the new naming options

tommyshem committed:
move all imports to the top of the file

tommyshem committed:
add comments and change parse_args to local
as only used in the plugin code

tommyshem committed:
update key bindings names as micro v2 uses lua
in front of the pluginname.function
eg lua:filemanager.toggle_tree

tommyshem commented:
I think this is ready for testing now.

sum01 commented:
So I just tested & it seems the width is no longer maintained at 30%.
When I open files it suddenly goes to 50%.

Side-note, seems `.steampath` is throwing an error when it tries to read it. On my system it's a symlink to an empty file. It's still putting the file in the tree, just spits out an error which might be worth investigating.

tommyshem commented:
micro v2 does not have `LockWidth=True` any more could not find a way of doing it.

Looked at symlink and it works on mac-os.

sum01 committed:
Fix for width getting messed up

This makes sure we re-size any time the width is messed with.
Since there's no width locking anymore, this is the best workaround I
could think up.

sum01 committed:
Use scratch type like before

Instead of a log type, switch back to the scratch like it was
previously.

sum01 committed:
Remove unnecessary extra install instruction

It's not needed to be in the main repo.

Also fixed some spacing with Prettier.

sum01 commented:
Tell me how these changes work out. From my testing it seems to do what I hoped it would.

sum01 committed:
Fix weird error on close of micro

It is calling the resize_everything function when the bufPane
is nil, even though it seemingly shouldn't, which was causing errors. So
this prevents errors on close.

sum01 commented:
`preMousePress` isn't triggering anymore, which breaks mouse support.

Not sure what to do about that. It seems it might not be working? https://gitter.im/zyedidia/micro?at=5ee7cee424a3382d5d6b7ab6

godalming123 commented:
why did you close this @tommyshem ?
